compiler/qsc_eval/src/intrinsic/utils.rs

```diff
@@ -0,0 +1,130 @@
+    for (curr_label, curr_val) in state_iter {
+        let dump_label = curr_label & &dump_mask;
+        let other_label = curr_label & &other_mask;
+
+        // If either the state identified by the dump mask or the state identified by the other mask
+        // is None, that means it has zero amplitude and we can conclude the state is not separable.
+        let Some(dump_val) = state.get(&(&dump_label | &other_base_label)) else {
+            return None;
+        };
+        let Some(other_val) = state.get(&(&dump_base_label | &other_label)) else {
+            return None;
+        };
+
+        if !(dump_val * other_val - base_val * curr_val)
+            .norm()
+            .is_nearly_zero()
+        {
+            // Coefficients are not equal, so the state is not separable.
+            return None;
+        }
+
+        if let Entry::Vacant(entry) = dump_state.entry(dump_label) {
+            let amplitude = curr_val / other_val;
+            let norm = amplitude.norm();
+            if !norm.is_nearly_zero() {
+                entry.insert(amplitude);
+                dump_norm += norm;
+            }
+        }
+        if let Entry::Vacant(entry) = other_state.entry(other_label) {
+            let amplitude = curr_val / dump_val;
+            let norm = amplitude.norm();
+            if !norm.is_nearly_zero() {
+                entry.insert(amplitude);
+            }
+        }
+    }
+
+    // If the product of the collected states is not equal to the total number of input states, then that
+    // implies some states are zero amplitude that would have to be non-zero for the state to be separable.
+    if state.len() != dump_state.len() * other_state.len() {
+        return None;
+    }
+
+    let dump_norm = 1.0 / dump_norm.sqrt();
+    Some(
+        dump_state
+            .into_iter()
+            .filter_map(|(label, val)| {
+                // Normalize the dump state by the collected factor.
+                let new_val = val * dump_norm;
+
+                // Drop any zero amplitude states.
+                if new_val.is_nearly_zero() {
+                    None
+                } else {
+                    // Reorder the bits in the label to match the provided qubit order.
+                    let mut new_label = BigUint::zero();
+                    for (i, q) in qubits.iter().enumerate() {
+                        // Note that the qubit order is reversed to match the order of the qubits in the state.
+                        if label.bit((qubit_count - *q - 1) as u64) {
+                            new_label.set_bit((qubits.len() - i - 1) as u64, true);
+                        }
+                    }
+
+                    Some((new_label, new_val))
+                }
+            })
+            .collect(),
+    )
+}
+
+trait NearlyZero {
+    fn is_nearly_zero(&self) -> bool;
+}
+
+impl NearlyZero for f64 {
+    fn is_nearly_zero(&self) -> bool {
+        self.max(0.0) - 0.0_f64.min(*self) <= 1e-10
```

Comment:
Minor suggestion, but was dealing with some math libs over the weekend and most commonly saw this done as something like `self.abs() <= 1e-10`. That might read a little easier and be less operations to perform.